### PR TITLE
Convert Catalog to Native JSON

### PR DESCRIPTION
**Pull request by dthain**
Not ready to merge yet.

Working:
- Catalog uses JSON natively.
- jx_database now keeps history as JSON diffs.

Still need to do:
- Update deltadb to use JSON natively.
- Remove unused parts of nvpair.

### CODE REVIEW

**Comment by dthain** [comment]
Ok, here it is:
- JX library now substantially complete.
- The catalog server uses JX natively, while still accepting/publishing nvpair reports.
- The catalog server uses the JX database to record its history.
- DeltaDB uses the JX database natively, all combined into one executable.
- Parts of nvpair printing/export that are duplicated by JX have been removed.
- catalog_update can now send arbitrary json records.
- Docs now include catalog_server, catalog_update, and deltadb_query explicitly.

Future work for someone else:
- Add support for arbitrary expression evaluation into JX.

@btovar please take a look

**Comment by dthain** [comment]
Whoops, build failed after rebase...

**Comment by dthain** [comment]
Now it's working.

File: deltadb/src/deltadb_expr.c [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+};
+
+struct deltadb_expr * deltadb_expr_create( const char *str, struct deltadb_expr *next )
+{
+	struct deltadb_expr *e = malloc(sizeof(*e));
+	e->param = strdup(str);
+	char *delim = strpbrk(e->param, "<>=!");
+	e->val = strpbrk(delim, "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ");
+	int operator_size = (int)(e->val-delim);
+	strncpy(e->operator,delim,operator_size);
```
Comment (btovar) [inline deltadb/src/deltadb_expr.c]: This is going to segfault if e->param does not contain an operator. (delim and e->val could be NULL).

File: deltadb/src/deltadb_expr.c [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	delim[0] = '\0';
+	e->next = next;
+	return e;
+}
+
+static int is_number(char const* p)
+{
+	char* end;
+	strtod(p, &end);
+	return !*end;
```
Comment (btovar) [inline deltadb/src/deltadb_expr.c]: It think here you want return end != p.

From the man page, if no conversion is performed, end is set to p.

File: deltadb/src/deltadb_query.c [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+}
+
+/*
+Play the log from starttime to stoptime by opening the appropriate
+checkpoint file and working ahead in the various log files.
+*/
+
+static int log_play_time( struct deltadb *db, time_t starttime, time_t stoptime )
+{
+	char filename[1024];
```
Comment (btovar) [inline deltadb/src/deltadb_query.c]: Suggest PATH_MAX instead.

File: deltadb/src/deltadb_query.c [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+
+	int year = starttm->tm_year + 1900;
+	int day = starttm->tm_yday;
+
+	struct tm *stoptm = localtime(&stoptime);
+
+	int stopyear = stoptm->tm_year + 1900;
+	int stopday = stoptm->tm_yday;
+
+	sprintf(filename,"%s/%d/%d.ckpt",db->logdir,year,day);
```
Comment (btovar) [inline deltadb/src/deltadb_query.c]: Suggest snprintf with PATH_MAX above.

File: deltadb/src/deltadb_query.c [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+	checkpoint_read(db,filename);
+
+	while(1) {
+		sprintf(filename,"%s/%d/%d.log",db->logdir,year,day);
```
Comment (btovar) [inline deltadb/src/deltadb_query.c]: Idem.

File: deltadb/src/deltadb_query.c [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+		}
+
+		day++;
+		if(day>=365) {
+			year++;
+			day = 0;
+		}
+
+		// If we have passed the file, stop.
+		if(year>=stopyear && day>stopday) break;
```
Comment (btovar) [inline deltadb/src/deltadb_query.c]: This will miss the last day of a leap year. Is it important?

**Comment by dthain** [comment]
All good points.  How does this look?

**Comment by btovar** [comment]
The changes look good.

**Pull request merged**